### PR TITLE
Feat/generate new key command state index issue

### PR DESCRIPTION
**Pull request by ognjenkurtic**
This PR implements execution of the generate new key circuit and smart contract method. As part of it, new verifying key is created for this proof and added to the the VKregistry contract as part of prepare_test.sh script.

Implementation can be verified by running the testVotingWithKeyDeactivation CLI test

### CODE REVIEW

**Review by grujicksenija — APPROVED** [review]

File: core/ts/MaciState.ts [hunk 1]
```diff
@@ -50,6 +49,12 @@ interface MaxValues {
     maxVoteOptions: number;
 }
 
+interface DeactivatedKeyEvent {
+    keyHash: BigInt;
+    c1: bigint[];
```
Comment (0xjei) [inline core/ts/MaciState.ts:54]: Change to `BigInt`

File: core/ts/MaciState.ts [hunk 2]
```diff
@@ -430,17 +440,21 @@ class Poll {
...
+                currentStateLeaves.push(this.stateLeaves[0].asCircuitInputs());
             }
 
             // Verify deactivation message
             const status = deactCommand.cmdType.toString() == '1' // Check message type
                 && computedStateIndex != -1
                 && signature != null
                 && verifySignature(
-                    deactMessage.hash(encPubKey),
+                    deactCommand.hash(),
```
Comment (0xjei) [inline core/ts/MaciState.ts:457]: `verifySignature` seems to take a message as first parameter. The command hash should be different from the signature. Could you please check that you are passing the correct value?

File: core/ts/MaciState.ts [hunk 3]
```diff
@@ -549,47 +560,42 @@ class Poll {
...
         }
 
-        const deactivatedKeyHash = hash3([...deactivatedPublicKey.asArray(), salt]);
-        const deactivatedKeyIndex = this.deactivatedKeyEvents.findIndex(d => d.keyHash === deactivatedKeyHash);
+        const deactivatedKeyHash: BigInt = hash3([...deactivatedPublicKey.asArray(), salt]);
+        const deactivatedKeyIndex = this.deactivatedKeyEvents.findIndex(d => d.keyHash.toString() == deactivatedKeyHash.toString());
 
         if (deactivatedKeyIndex === -1) {
-            console.log("Key index is -1");
+            console.log("Deactivated key is missing from the deactivated key events collection from contract");
```
Comment (0xjei) [inline core/ts/MaciState.ts:567]: +1

File: core/ts/MaciState.ts [hunk 4]
```diff
@@ -549,47 +560,42 @@ class Poll {
...
 
-        const stateIndexInt = parseInt(stateIndex.toString());
-        const computedStateIndex = stateIndexInt > 0 && stateIndexInt <= this.numSignUps ? stateIndexInt - 1 : -1;
-        const currentStateLeaves = [this.stateLeaves[computedStateIndex]];
-
         const z = BigInt(42);
 
         const [c1r, c2r] = elGamalRerandomize(
-            newPublicKey.rawPubKey,
+            this.coordinatorKeypair.pubKey.rawPubKey,
```
Comment (0xjei) [inline core/ts/MaciState.ts:576]: Should be `this.coordinatorKeypair.pubKey` (?)

Could you please check that you are passing the correct value?

File: core/ts/MaciState.ts [hunk 5]
```diff
@@ -549,47 +560,42 @@ class Poll {
...
             deactivatedKeyEvent.c2,
         );
 
-        // TODO: check if asCircuitInputs should be used
         const nullifier = hash2([BigInt(deactivatedPrivateKey.asCircuitInputs()), salt]);
 
-        // TODO: set proper value for voiceCreditBalance
-        const voiceCreditBalance = BigInt(100)
+        // TODO: set proper value for newCreditBalance - take from input param?
+        const newCreditBalance = BigInt(5)
```
Comment (0xjei) [inline core/ts/MaciState.ts:585]: Why the `newCreditBalance` is set to 5?

File: core/ts/MaciState.ts [hunk 6]
```diff
@@ -601,8 +607,6 @@ class Poll {
             deactivatedKeyEvent.c1,
             deactivatedKeyEvent.c2
         )
```
Comment (0xjei) [inline core/ts/MaciState.ts:609]: Why we don't return the `circuitInputs` along with the `kcommand` anymore?

**Review by 0xjei — COMMENTED** [review]
Great work, I have left a few comments!

File: core/ts/MaciState.ts [hunk 7]
```diff
@@ -1876,6 +1880,7 @@ const genProcessVkSig = (
         BigInt(_voteOptionTreeDepth)
 }
 
+// TODO: Not used anywhere.Also one for new key generation is missing.
 const genDeactivationVkSig = (
     _messageQueueSize: number,
     _stateTreeDepth: number,
```
Comment (0xjei) [inline core/ts/MaciState.ts:1886]: `genNewKeyGenerationVkSig` is missing in `MaciState`

File: core/ts/MaciState.ts (hunk 6 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:609]: We are returning const { circuitInputs, encPubKey, message }  from the call directly to the caller.

File: core/ts/MaciState.ts (hunk 5 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:585]: Added a more elaborate TODO comment

File: contracts/hardhat.config.js [hunk 8]
```diff
@@ -21,7 +21,7 @@ module.exports = {
 				mnemonic:
 					'candy maple cake sugar pudding cream honey rich smooth crumble sweet treat',
 			},
-			loggingEnabled: false,
+			loggingEnabled: true,
```
[the remark this answers is not in the document]
Comment (ognjenkurtic) [reply contracts/hardhat.config.js:24]: Oversight.Fixed

File: core/ts/MaciState.ts (hunk 1 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:54]: Fixed

File: core/ts/MaciState.ts (hunk 4 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:576]: this.coordinatorKeypair.pubKey is a class PubKey from DomainObejcts.
elGamalRerandomize expects type PubKey frm Crypto, which is just a BigInt[] and which coresponds to the this.coordinatorKeypair.pubKey.rawPubKey

File: core/ts/MaciState.ts (hunk 7 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:1886]: Yes, i left that in the TODO comment because i was not sure do we need these as the genDeactivationVkSig is not used anywhere. Let's resolve once Aleksandar is back,

File: core/ts/MaciState.ts (hunk 2 above)
Comment (ognjenkurtic) [reply core/ts/MaciState.ts:457]: msg param from VerifySignature is just a BigInt, so a hash is valid. Old param was also just a hash. This change was introduced by Aleksandar and i believe was changed because the previous one was either unnecessary or plain wrong. Tests pass so we can verify with Aleksandar once he is back,

Comment (0xjei) [reply core/ts/MaciState.ts:457]: Let me rephrase the question: I was wondering since `verifySignature` takes `message` as its first parameter - how come we're passing a `command` hash. Is this to be consider correct or not? If you'd like, we could punt back to when Alex is back

File: core/ts/MaciState.ts (hunk 4 above)
Comment (0xjei) [reply core/ts/MaciState.ts:576]: > this.coordinatorKeypair.pubKey is a class PubKey from DomainObejcts. elGamalRerandomize expects type PubKey frm Crypto, which is just a BigInt[] and which coresponds to the this.coordinatorKeypair.pubKey.rawPubKey

Thanks for the explanation! :+1:

File: core/ts/MaciState.ts (hunk 5 above)
Comment (0xjei) [reply core/ts/MaciState.ts:586]: Ahh, I see, we need to take it as new input parameter, makes sense now. What are the barriers to implementing it right now?

File: core/ts/MaciState.ts (hunk 7 above)
Comment (0xjei) [reply core/ts/MaciState.ts:1887]: Correct, but right now it should serve to initialize the VkRegistry contract with the key also of the `GenerateKeyFromDeactivated` circuit.
